[addons] Bring back audioengine addon interface (not ADSP)

AlwinEsch opened the pull request:
## Description
This is a mandatory for Webbrowser addon system!

There are no changes on Audio Engine and from usage mostly equal to other
audioengine calls in Kodi.

All addons not need a recompile with them, on `AddonToKodiFuncTable_Addon` the `AddonToKodiFuncTable_kodi_audioengine* kodi_audioengine;` was still defined.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
Tested on Webbrowser addon system
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thebrid commented on xbmc/addons/interfaces/AudioEngine.cpp:
```diff
@@ -0,0 +1,383 @@
+
+void Interface_AudioEngine::Init(AddonGlobalInterface* addonInterface)
+{
+  addonInterface->toKodi->kodi_audioengine = (AddonToKodiFuncTable_kodi_audioengine*)malloc(sizeof(AddonToKodiFuncTable_kodi_audioengine));
```
Why `malloc` rather than `new`?

AlwinEsch replied:
The generated array is for a `C` code table, I'm not sure that a `new` outlets with pure "C" code might not generate traceable errors. Think to 98% the `new` works also always but there are 2% left ;-).

But good for the note, have forgot to change them to `static_cast<AddonToKodiFuncTable_kodi_audioengine*>(...)`.

thebrid commented on xbmc/addons/interfaces/AudioEngine.cpp:
```diff
@@ -0,0 +1,383 @@
+void Interface_AudioEngine::DeInit(AddonGlobalInterface* addonInterface)
+{
+  if (addonInterface->toKodi && /* <-- Safe check, needed so long old addon way is present */
+      addonInterface->toKodi->kodi_audioengine)
```
The second check here is redundant.

AlwinEsch replied:
This has created here on Linux crashes, where on PVR and Game addons (still in old interfaces) and `addonInterface->toKodi` not becomes set (from here https://github.com/xbmc/xbmc/blob/master/xbmc/addons/binary-addons/AddonDll.cpp#L184).

On new becomes this used: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/binary-addons/AddonDll.cpp#L264

AlwinEsch replied:
Maybe I understand wrong and you mean to leave `addonInterface->toKodi->kodi_audioengine` away? And let the check to `free` itself?

thebrid replied:
@AlwinEsch yes, the latter. `free()` (or `delete`) will do nothing on a null pointer. `clang-tidy` [complains](https://clang.llvm.org/extra/clang-tidy/checks/readability-delete-null-pointer.html) if we check this. :smile:

AlwinEsch replied:
Have removed the `addonInterface->toKodi->kodi_audioengine` check

AlwinEsch committed:
[addons] Bring back audioengine addon interface (not ADSP)

This is a mandatory for Webbrowser addon system!

There are no changes on Audio Engine and from usage mostly equal to other
audioengine calls in Kodi.
